File: packages/oxlint-plugin-react-doctor/src/plugin/rules/jotai/jotai-derived-atom-returns-fresh-object.ts
import { defineRule } from "../../utils/define-rule.js";
import {
  isImportedFromModule,
  getImportedNameFromModule,
} from "../../utils/find-import-source-for-name.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import { walkAst } from "../../utils/walk-ast.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

// HACK: jotai propagates derived-atom values with `Object.is`. There
// is no shallow compare and no custom equality on a plain
// `atom((get) => ...)`. A derivation that returns a fresh
// ObjectExpression / ArrayExpression literal — OR a method-chain that
// produces a fresh array / object on every read — fails Object.is on
// every notify and re-renders every consumer, even when the field
// values didn't change. Measured: a fresh-object derivation committed
// 44× more than the equivalent `useQuery` on no-op refetches.
// Fix: split into multiple primitive derived atoms (each `Object.is`-
// dedup-able), or wrap with `selectAtom(source, fn, shallow)` from
// `jotai/utils` if a single wrapper object is genuinely required.

type FunctionExpressionLike =
  | EsTreeNodeOfType<"ArrowFunctionExpression">
  | EsTreeNodeOfType<"FunctionExpression">;

const isAtomFromJotai = (callExpression: EsTreeNodeOfType<"CallExpression">): boolean => {
  if (!isNodeOfType(callExpression.callee, "Identifier")) return false;
  const localName = callExpression.callee.name;
  if (!isImportedFromModule(callExpression, localName, "jotai")) return false;
  return getImportedNameFromModule(callExpression, localName, "jotai") === "atom";
};

const isFunctionLike = (node: EsTreeNode | null | undefined): node is FunctionExpressionLike =>
  Boolean(
    node &&
    (isNodeOfType(node, "ArrowFunctionExpression") || isNodeOfType(node, "FunctionExpression")),
  );

const getFirstParameterName = (fn: FunctionExpressionLike): string | null => {
  const parameters = fn.params ?? [];
  if (parameters.length !== 1) return null;
  const first = parameters[0];
  return isNodeOfType(first, "Identifier") ? first.name : null;
};

// Array-prototype methods that always allocate a fresh array. Includes
// the immutable variants from ES2023 (`toSorted`, `toReversed`,
// `toSpliced`, `with`) plus the classic Array.prototype mutators that
// return the modified-but-fresh-shape array (`.sort()`, `.reverse()`
// — note these mutate AND return the receiver; if the receiver is a
// fresh array from `.map()` upstream, the chain still allocates).
const FRESH_ARRAY_INSTANCE_METHODS = new Set([
  "filter",
  "map",
  "flatMap",
  "slice",
  "concat",
  "flat",
  "toSorted",
  "toReversed",
  "toSpliced",
  "with",
  "sort",
  "reverse",
]);

// Static methods that allocate a fresh array / object from upstream.
const FRESH_STATIC_OBJECT_CALLS: Record<string, ReadonlySet<string>> = {
  Object: new Set(["keys", "values", "entries", "fromEntries", "assign", "create"]),
  Array: new Set(["from", "of"]),
};

interface FreshReturn {
  kind: "object" | "array";
  reportNode: EsTreeNode;
}

const freshFromObjectLiteral = (expression: EsTreeNode): FreshReturn | null => {
  if (isNodeOfType(expression, "ObjectExpression")) {
    return { kind: "object", reportNode: expression };
  }
  if (isNodeOfType(expression, "ArrayExpression")) {
    return { kind: "array", reportNode: expression };
  }
  return null;
};

// Only the OUTERMOST method in the chain decides whether the atom's
// value is a fresh structure. Walking inward past non-matching methods
// produces false positives: `get(users).filter(fn).reduce(sum, 0)`
// consumes the filtered array and returns a primitive that dedupes
// via Object.is correctly — even though `.filter()` appears inside
// the chain.
const freshFromMethodChain = (expression: EsTreeNode): FreshReturn | null => {
  if (!isNodeOfType(expression, "CallExpression")) return null;
  const callee = expression.callee;
  if (!isNodeOfType(callee, "MemberExpression")) return null;
  if (callee.computed) return null;
  if (!isNodeOfType(callee.property, "Identifier")) return null;
  const methodName = callee.property.name;
  if (FRESH_ARRAY_INSTANCE_METHODS.has(methodName)) {
    return { kind: "array", reportNode: expression };
  }
  // Static-method form: `Object.entries(get(x))`, `Array.from(get(x))`.
  if (isNodeOfType(callee.object, "Identifier")) {
    const staticMethods = FRESH_STATIC_OBJECT_CALLS[callee.object.name];
    if (staticMethods?.has(methodName)) {
      return {
        kind:
          callee.object.name === "Array" ||
          methodName === "keys" ||
          methodName === "values" ||
          methodName === "entries"
            ? "array"
            : "object",
        reportNode: expression,
      };
    }
  }
  return null;
};

const classifyReturnedExpression = (
  expression: EsTreeNode | null | undefined,
): FreshReturn | null => {
  if (!expression) return null;
  const inner = stripParenExpression(expression);
  const literalReturn = freshFromObjectLiteral(inner);
  if (literalReturn) return literalReturn;
  return freshFromMethodChain(inner);
};

const collectTopLevelReturnExpressions = (
  block: EsTreeNodeOfType<"BlockStatement">,
): Array<EsTreeNode | null | undefined> => {
  const returns: Array<EsTreeNode | null | undefined> = [];
  walkAst(block, (child) => {
    if (
      isNodeOfType(child, "FunctionDeclaration") ||
      isNodeOfType(child, "FunctionExpression") ||
      isNodeOfType(child, "ArrowFunctionExpression")
    ) {
      // Don't descend into nested functions — their returns belong to
      // their own control-flow scope.
      return false;
    }
    if (isNodeOfType(child, "ReturnStatement")) returns.push(child.argument);
  });
  return returns;
};

const getFreshReturnForFunction = (fn: FunctionExpressionLike): FreshReturn | null => {
  const body = fn.body;
  if (!body) return null;
  // Concise arrow body: `(get) => ({ ... })` / `(get) => [...]` / `(get) => get(x).filter(...)`.
  if (!isNodeOfType(body, "BlockStatement")) {
    return classifyReturnedExpression(body);
  }
  // Block body: every reachable return must produce a fresh structure
  // for the diagnostic to hold across every notify. If ANY return
  // emits a primitive / `get(...)` member chain, the atom can still
  // dedupe on some paths and the recommendation no longer fits.
  const returnExpressions = collectTopLevelReturnExpressions(body);
  if (returnExpressions.length === 0) return null;
  let firstFresh: FreshReturn | null = null;
  for (const returnArgument of returnExpressions) {
    const classification = classifyReturnedExpression(returnArgument);
    if (!classification) return null;
    if (!firstFresh) firstFresh = classification;
  }
  return firstFresh;
};

const functionBodyReferencesGetParameter = (
  fn: FunctionExpressionLike,
  getParameterName: string,
): boolean => {
  const body = fn.body;
  if (!body) return false;
  let found = false;
  walkAst(body, (child) => {
    if (found) return false;
    // Don't descend into nested functions — their `get(...)` calls
    // belong to their own closure and don't prove the outer atom
    // reads from upstream.
    if (
      isNodeOfType(child, "FunctionDeclaration") ||
      isNodeOfType(child, "FunctionExpression") ||
      isNodeOfType(child, "ArrowFunctionExpression")
    ) {
      if (child !== fn) return false;
    }
    if (!isNodeOfType(child, "CallExpression")) return;
    if (!isNodeOfType(child.callee, "Identifier")) return;
    if (child.callee.name === getParameterName) {
      found = true;
      return false;
    }
  });
  return found;
};

export const jotaiDerivedAtomReturnsFreshObject = defineRule<Rule>({
  id: "jotai-derived-atom-returns-fresh-object",
  title: "Derived atom returns fresh object",
  severity: "warn",
  recommendation:
    "Split it into one derived atom per field (each compares cleanly with Object.is), or wrap with `selectAtom(source, fn, shallow)` from jotai/utils if you really need one object.",
  create: (context: RuleContext) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      if (!isAtomFromJotai(node)) return;
      const args = node.arguments ?? [];
      if (args.length === 0) return;
      const reader = args[0];
      if (!isFunctionLike(reader)) return;
      // Write-only / read-write atoms have a second `set` parameter
      // and produce a different propagation shape. Out of v1 scope.
      const getParameterName = getFirstParameterName(reader);
      if (!getParameterName) return;

      const freshReturn = getFreshReturnForFunction(reader);
      if (!freshReturn) return;
      // A function body that never reads upstream via `get(...)` isn't
      // a derived atom — it's a constant atom the author wrote with
      // a function. The fresh-literal cost only matters when the
      // upstream is actually being propagated.
      if (!functionBodyReferencesGetParameter(reader, getParameterName)) return;

      const shape = freshReturn.kind === "object" ? "object" : "array";
      context.report({
        node: freshReturn.reportNode,
        message: `This derived atom returns a new ${shape} each time, so jotai's Object.is check fails & re-renders every consumer on every update. Split into one atom per field, or use \`selectAtom(source, fn, shallow)\`.`,
      });
    },
  }),
});

File: native/oxlint/rules/jotai-derived-atom-returns-fresh-object.rs
use oxc_ast::{
    AstKind,
    ast::{Argument, Expression},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::{GetSpan, Span};

use crate::{AstNode, context::LintContext, rule::Rule};

const FRESH_ARRAY_METHODS: [&str; 12] = [
    "filter",
    "map",
    "flatMap",
    "slice",
    "concat",
    "flat",
    "toSorted",
    "toReversed",
    "toSpliced",
    "with",
    "sort",
    "reverse",
];

#[derive(Debug, Default, Clone)]
pub struct JotaiDerivedAtomReturnsFreshObject;

declare_oxc_lint!(
    /// Warn when a derived Jotai atom returns a fresh object or array.
    JotaiDerivedAtomReturnsFreshObject,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Derived atom returns fresh object.",
);

impl Rule for JotaiDerivedAtomReturnsFreshObject {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(atom_call) = node.kind() else {
            return;
        };
        let Expression::Identifier(callee) = &atom_call.callee else {
            return;
        };
        let Some(import_entry) = resolve_identifier_import(callee, ctx) else {
            return;
        };
        if import_entry.module_request.name() != "jotai"
            || !matches!(
                &import_entry.import_name,
                crate::module_record::ImportImportName::Name(imported_name)
                    if imported_name.name() == "atom"
            )
        {
            return;
        }
        let Some(reader) = atom_call
            .arguments
            .first()
            .and_then(Argument::as_expression)
        else {
            return;
        };
        let (function_node_id, function_span, return_expression, get_parameter_symbol_id) =
            match reader {
                Expression::ArrowFunctionExpression(function) => {
                    if function.params.items.len() != 1 {
                        return;
                    }
                    let Some(parameter) = function.params.items[0].pattern.get_binding_identifier()
                    else {
                        return;
                    };
                    (
                        function.node_id.get(),
                        function.span,
                        function.get_expression(),
                        parameter.symbol_id(),
                    )
                }
                Expression::FunctionExpression(function) => {
                    if function.params.items.len() != 1 {
                        return;
                    }
                    let Some(parameter) = function.params.items[0].pattern.get_binding_identifier()
                    else {
                        return;
                    };
                    (
                        function.node_id.get(),
                        function.span,
                        None,
                        parameter.symbol_id(),
                    )
                }
                _ => return,
            };
        if !jotai_reader_calls_get(function_node_id, get_parameter_symbol_id, ctx) {
            return;
        }
        if let Some(expression) = return_expression {
            if let Some((shape, span)) = jotai_fresh_return(expression) {
                jotai_report_fresh_return(shape, span, ctx);
            }
            return;
        }
        let returns = ctx
            .nodes()
            .iter()
            .filter_map(|candidate| {
                let AstKind::ReturnStatement(statement) = candidate.kind() else {
                    return None;
                };
                if !function_span.contains_inclusive(statement.span)
                    || nearest_jotai_function_node_id(candidate, ctx) != Some(function_node_id)
                {
                    return None;
                }
                Some(statement.argument.as_ref().and_then(jotai_fresh_return))
            })
            .collect::<Vec<_>>();
        if returns.is_empty() || returns.iter().any(Option::is_none) {
            return;
        }
        if let Some((shape, span)) = returns.into_iter().flatten().next() {
            jotai_report_fresh_return(shape, span, ctx);
        }
    }
}

fn jotai_reader_calls_get(
    function_node_id: oxc_semantic::NodeId,
    parameter_symbol_id: oxc_semantic::SymbolId,
    ctx: &LintContext<'_>,
) -> bool {
    ctx.scoping()
        .get_resolved_references(parameter_symbol_id)
        .any(|reference| {
            let reference_node = ctx.nodes().get_node(reference.node_id());
            let parent = ctx.nodes().parent_node(reference_node.id());
            let AstKind::CallExpression(call) = parent.kind() else {
                return false;
            };
            call.callee.span() == reference_node.span()
                && nearest_jotai_function_node_id(reference_node, ctx) == Some(function_node_id)
        })
}

fn nearest_jotai_function_node_id(
    node: &AstNode<'_>,
    ctx: &LintContext<'_>,
) -> Option<oxc_semantic::NodeId> {
    ctx.nodes().ancestors(node.id()).find_map(|ancestor| {
        matches!(
            ancestor.kind(),
            AstKind::Function(_) | AstKind::ArrowFunctionExpression(_)
        )
        .then(|| ancestor.id())
    })
}

fn jotai_fresh_return(expression: &Expression<'_>) -> Option<(&'static str, Span)> {
    let expression = expression.get_inner_expression();
    match expression {
        Expression::ObjectExpression(object) => Some(("object", object.span)),
        Expression::ArrayExpression(array) => Some(("array", array.span)),
        Expression::CallExpression(call) => {
            let oxc_ast::ast::MemberExpression::StaticMemberExpression(member) =
                call.callee.as_member_expression()?
            else {
                return None;
            };
            let method_name = member.property.name.as_str();
            let receiver = member.object.get_inner_expression();
            if matches!(method_name, "sort" | "reverse") && jotai_fresh_return(receiver).is_none() {
                return None;
            }
            if FRESH_ARRAY_METHODS.contains(&method_name) {
                return Some(("array", call.span));
            }
            let Expression::Identifier(namespace) = receiver else {
                return None;
            };
            match (namespace.name.as_str(), method_name) {
                ("Array", "from" | "of") => Some(("array", call.span)),
                ("Object", "keys" | "values" | "entries") => Some(("array", call.span)),
                ("Object", "fromEntries" | "create") => Some(("object", call.span)),
                ("Object", "assign") => call
                    .arguments
                    .first()
                    .and_then(Argument::as_expression)
                    .is_some_and(|target| {
                        matches!(
                            target.get_inner_expression(),
                            Expression::ObjectExpression(_) | Expression::ArrayExpression(_)
                        )
                    })
                    .then_some(("object", call.span)),
                _ => None,
            }
        }
        _ => None,
    }
}

fn jotai_report_fresh_return(shape: &str, span: Span, ctx: &LintContext<'_>) {
    ctx.diagnostic(
        OxcDiagnostic::warn(format!(
            "This derived atom returns a new {shape} each time, so jotai's Object.is check fails & re-renders every consumer on every update. Split into one atom per field, or use `selectAtom(source, fn, shallow)`."
        ))
        .with_label(span),
    );
}
